text-serializer-ansi/src/main/java/net/kyori/adventure/text/serializer/ansi/ANSIComponentSerializerImpl.java

```diff
@@ -0,0 +1,276 @@
+   */
+  @Override
+  public @NonNull Component deserialize(final @NonNull String input) {
+    final Matcher matcher = STRIP_ESC_CODES.matcher(input);
+    String out = input;
+    while(matcher.find()){
+      out = out.replace(matcher.group(), "");
+    }
+    return Component.text(out);
+  }
+
+  /**
+   * Serializes a component into an output of type {@link ANSIComponentSerializer}.
+   *
+   * @param component the component
+   * @return the output
+   */
+  @Override
+  public @NonNull String serialize(final @NonNull Component component){
+    final Porridge state = new Porridge();
+    state.append(component);
+    return state.toString();
+  }
+
+  @Override
+  public @NonNull Builder toBuilder(){
+    return new BuilderImpl(this);
+  }
+
+  static final class BuilderImpl implements Builder {
+    private boolean colorDownSample = true;
+
+    BuilderImpl(){
+
+    }
+
+    BuilderImpl(final @NonNull ANSIComponentSerializerImpl serializer){
+      this.colorDownSample = serializer.colorDownSample;
+    }
+
+    @Override
+    public @NonNull Builder downSample(final boolean downSample) {
+      this.colorDownSample = downSample;
+      return this;
+    }
+
+    @Override
+    public @NonNull ANSIComponentSerializer build() {
+      return new ANSIComponentSerializerImpl(this.colorDownSample);
+    }
+  }
+
+  private final class Porridge {
+    private final Ansi ansi = Ansi.ansi();
+    private final Style style = new Style();
+
+    @Override
+    public String toString(){
+      return this.ansi.toString();
+    }
+
+    void append(final @NonNull Component component){
+      this.append(component, new Style());
+    }
+
+    private void append(final @NonNull Component component, final @NonNull Style style){
+      style.apply(component);
+
+      if(component instanceof TextComponent){
+        final String content = ((TextComponent) component).content();
+        if(!content.isEmpty()){
+          style.applyFormat();
+          this.ansi.a(content);
+          this.ansi.a(Formats.SOFT_RESET);
+        }
+      }
+      if(component instanceof TranslatableComponent){
+        final Component c = TranslatableComponentRenderer.get().render(component, Locale.getDefault());
+        if(c instanceof TextComponent){
+          final String translatedContent = ((TextComponent) c).content();
+          if(!translatedContent.isEmpty()){
+            style.applyFormat();
+            this.ansi.a(translatedContent);
+            this.ansi.a(Formats.SOFT_RESET);
+          }
+        }
+      }
+      final List<Component> children = component.children();
+      if(!children.isEmpty()){
+        final Style childrenStyle = new Style(style);
+        for(final Component child : children){
+          this.append(child, childrenStyle);
+          childrenStyle.set(style);
+        }
+      }
+    }
+
+    void append(final @NonNull TextFormat format) {
+      final Ansi a = ANSIComponentSerializerImpl.this.toANSI(format);
```

Comment:
```suggestion
      final Ansi ansi = ANSIComponentSerializerImpl.this.toANSI(format);
```

Avoid single name variables.

Reply:
I dislike naming private method variables that will cause a class field to be shadowed.  I think it can cause confusion as a result I thought `output `